model/lifecycle.go

```diff
@@ -57,25 +57,85 @@ type VersionToRestart struct {
-
-	// Update activation for all builds before updating their tasks so the version won't spend
-	// time in an intermediate state where only some builds are updated
-	if err = build.UpdateActivation(buildIDs, active, caller); err != nil {
-		return errors.Wrapf(err, "setting activation for builds in version '%s'", versionId)
+	if err = version.SetActivated(active); err != nil {
+		return errors.Wrapf(err, "setting activated for version '%s'", versionId)
+	}
+	changed, err := setTaskActivationForVersion(versionId, active, false, nil, caller)
+	if err != nil {
+		return errors.Wrapf(err, "setting activation for tasks in version '%s'", versionId)
+	}
+	if active && changed && evergreen.IsFinishedVersionStatus(version.Status) {
+		if err = version.UpdateStatus(evergreen.VersionCreated); err != nil {
+			return errors.Wrapf(err, "setting status for version '%s'", versionId)
+		}
 	}
+	return nil
+}
 
-	return errors.Wrapf(setTaskActivationForBuilds(buildIDs, active, false, nil, caller),
-		"setting activation for tasks in version '%s'", versionId)
+func setTaskActivationForVersion(versionId string, active, withDependencies bool, ignoreTasks []string, caller string) (bool, error) {
+	// If activating a task, set the ActivatedBy field to be the caller
+	if active {
+		q := bson.M{
+			task.VersionKey: versionId,
+			task.StatusKey:  evergreen.TaskUndispatched,
+		}
+		if len(ignoreTasks) > 0 {
+			q[task.IdKey] = bson.M{"$nin": ignoreTasks}
+		}
+		tasksToActivate, err := task.FindAll(db.Query(q).WithFields(task.IdKey, task.DependsOnKey, task.ExecutionKey, task.BuildIdKey))
+		if err != nil {
+			return false, errors.Wrap(err, "getting tasks to activate")
+		}
+		if withDependencies {
+			dependOn, err := task.GetRecursiveDependenciesUp(tasksToActivate, nil)
+			if err != nil {
+				return false, errors.Wrap(err, "getting recursive dependencies")
+			}
+			tasksToActivate = append(tasksToActivate, dependOn...)
+		}
+		if len(tasksToActivate) > 0 {
+			var buildIds []string
+			for _, task := range tasksToActivate {
+				if !utility.StringSliceContains(buildIds, task.BuildId) {
+					buildIds = append(buildIds, task.BuildId)
+				}
+			}
+			if err = build.UpdateActivationAndStatus(buildIds, active, caller); err != nil {
+				return false, errors.Wrapf(err, "setting activation for builds in version '%s'", versionId)
+			}
+			if err = task.ActivateTasks(tasksToActivate, time.Now(), withDependencies, caller); err != nil {
+				return false, errors.Wrap(err, "updating tasks for activation")
+			}
+			return true, nil
+		}
+	} else {
+		query := bson.M{
+			task.VersionKey: versionId,
+			task.StatusKey:  evergreen.TaskUndispatched,
+		}
+		// if the caller is the default task activator only deactivate tasks that have not been activated by a user
+		if evergreen.IsSystemActivator(caller) {
+			query[task.ActivatedByKey] = bson.M{"$in": evergreen.SystemActivators}
+		}
+
+		tasks, err := task.FindAll(db.Query(query).WithFields(task.IdKey, task.ExecutionKey))
```

Comment:
Similarly, I don't think we need to query the tasks to do this update, if we refactor deactivateTasks to take in a query